File: Scr_Java/estructuras/ColaEnemigos.java
package estructuras;

import java.util.ArrayDeque;
import java.util.Queue;

import gestor.GestorJuego;
import modelo.Coordenada;
import modelo.Cozy;
import modelo.Torre;

// Cola circular: cada Cozy vivo que no llega al final se vuelve a encolar
// al final de la cola, simulando su avance continuo por la ruta.
public class ColaEnemigos {
    private static final int DANO_POR_FUGA = 1; // vida que pierde el jugador por cada fuga

    private final Queue<Cozy> colaCircular;

    public ColaEnemigos() {
        this.colaCircular = new ArrayDeque<>();
    }

    public void encolar(Cozy c) {
        colaCircular.add(c);
    }

    public void actualizarQuantum(Mapa mapa, GestorJuego estado) {
        // Se fija el tamanio ANTES del bucle: cada Cozy se procesa una sola vez
        // por quantum, aunque se vuelva a encolar dentro del mismo bucle.
        int cantidadAProcesar = colaCircular.size();
        Coordenada[] ruta = mapa.getRuta();
        Torre[] torresActivas = mapa.getTorresActivas();

        for (int i = 0; i < cantidadAProcesar; i++) {
            Cozy c = colaCircular.poll();
            Coordenada posActual = ruta[c.getIndiceRuta()];

            // Todas las torres en rango disparan sobre el enemigo en su posicion actual.
            for (Torre t : torresActivas) {
                if (c.estaVivo() && t.estaEnRango(posActual)) {
                    t.atacar(c);
                }
            }

            if (!c.estaVivo()) {
                estado.sumarPuntaje(c.getRecompensa());
                continue; // muere: se retira, no se re-encola
            }

            c.avanzar();
            if (c.llegoAlFinal(ruta.length)) {
                estado.restarVida(DANO_POR_FUGA);
                continue; // se fuga: se retira, no se re-encola
            }

            colaCircular.add(c); // sigue vivo en ruta: vuelve al final de la cola
        }
    }

    public boolean estaVacia() {
        return colaCircular.isEmpty();
    }

    public int tamanio() {
        return colaCircular.size();
    }
}

File: cpp/estructuras/ColaEnemigos.cpp
#include "ColaEnemigos.h"

ColaEnemigos::ColaEnemigos()
    : colaCircular() {
}

void ColaEnemigos::encolar(const Cozy& enemigo) {
    colaCircular.push(enemigo);
}

void ColaEnemigos::actualizarQuantum(Mapa& mapa, GestorJuego& estado) {

    // Se obtiene el tamanio ANTES de comenzar el bucle.
    // Esto garantiza que cada Cozy sea procesado una sola vez
    // por quantum, aunque posteriormente vuelva a ingresar a la cola.
    int cantidadAProcesar = static_cast<int>(colaCircular.size());

    const std::vector<Coordenada>& ruta = mapa.getRuta();
    std::vector<Torre*> torresActivas = mapa.getTorresActivas();

    for (int i = 0; i < cantidadAProcesar; i++) {

        // front() obtiene el primer elemento de la cola
        // y pop() lo elimina de la cola.
        Cozy enemigo = colaCircular.front();
        colaCircular.pop();

        // Se obtiene la posicion actual del enemigo dentro de la ruta.
        const Coordenada& posActual =
            ruta[enemigo.getIndiceRuta()];

        // Todas las torres activas comprueban si pueden atacar
        // al enemigo ubicado en su posicion actual.
        for (Torre* torre : torresActivas) {

            if (enemigo.estaVivo() &&
                torre->estaEnRango(posActual)) {

                // La torre aplica su ataque sobre la copia local del
                // enemigo (la misma que luego se vuelve a encolar).
                torre->atacar(enemigo);
            }
        }

        // Si el enemigo fue eliminado durante el ataque,
        // se suma su recompensa y no vuelve a la cola.
        if (!enemigo.estaVivo()) {
            estado.sumarPuntaje(enemigo.getRecompensa());
            continue;
        }

        // El enemigo avanza dentro de la ruta.
        enemigo.avanzar();

        // Si alcanza la posicion final, se considera una fuga.
        if (enemigo.llegoAlFinal(
                static_cast<int>(ruta.size()))) {

            estado.restarVida(DANO_POR_FUGA);
            continue;
        }

        // Si continua vivo y no ha llegado al final,
        // vuelve a colocarse al final de la cola.
        colaCircular.push(enemigo);
    }
}

bool ColaEnemigos::estaVacia() const {
    return colaCircular.empty();
}

int ColaEnemigos::tamanio() const {
    return static_cast<int>(colaCircular.size());
}
